src/Sylius/Bundle/CoreBundle/Processor/DummyCatalogPromotionProcessor.php

```diff
@@ -15,43 +15,35 @@
 
 use Doctrine\ORM\EntityManagerInterface;
 use Sylius\Bundle\CoreBundle\Applicator\CatalogPromotionApplicatorInterface;
+use Sylius\Component\Core\Model\CatalogPromotionInterface;
 use Sylius\Component\Core\Model\ProductInterface;
-use Sylius\Component\Core\Model\TaxonInterface;
-use Sylius\Component\Core\Repository\ProductRepositoryInterface;
-use Sylius\Component\Promotion\Model\CatalogPromotionInterface;
-use Sylius\Component\Taxonomy\Repository\TaxonRepositoryInterface;
+use Sylius\Component\Core\Provider\CatalogPromotionProductsProviderInterface;
 
 final class DummyCatalogPromotionProcessor implements CatalogPromotionProcessorInterface
 {
-    private ProductRepositoryInterface $productRepository;
-
-    private TaxonRepositoryInterface $taxonRepository;
+    private CatalogPromotionProductsProviderInterface $catalogPromotionProductsProvider;
 
     private CatalogPromotionApplicatorInterface $catalogPromotionApplicator;
 
     private EntityManagerInterface $entityManager;
 
     public function __construct(
-        ProductRepositoryInterface $productRepository,
-        TaxonRepositoryInterface $taxonRepository,
+        CatalogPromotionProductsProviderInterface $catalogPromotionProductsProvider,
         CatalogPromotionApplicatorInterface $catalogPromotionApplicator,
         EntityManagerInterface $entityManager
     ) {
-        $this->productRepository = $productRepository;
-        $this->taxonRepository = $taxonRepository;
+        $this->catalogPromotionProductsProvider = $catalogPromotionProductsProvider;
         $this->catalogPromotionApplicator = $catalogPromotionApplicator;
         $this->entityManager = $entityManager;
     }
 
     public function process(CatalogPromotionInterface $catalogPromotion): void
     {
-        /** @var TaxonInterface|null $taxon */
-        $taxon = $this->taxonRepository->findOneBy(['code' => 't_shirts']);
-        if ($taxon === null) {
+        $products = $this->catalogPromotionProductsProvider->provideEligibleProducts($catalogPromotion);
```

Comment:
We should apply discounts on variants, not products as the one existing rule bases on variants